### PR TITLE
docs(design): Add design doc

### PR DESCRIPTION
**Pull request by gaocegege**
Signed-off-by: Ce Gao <gaoce@caicloud.io>

### CODE REVIEW

File: docs/design_doc.md [hunk 1]
```diff
@@ -0,0 +1,261 @@
+## Kubeflow Controller Design Document
+
+## Overview
+
+Kubeflow controller is similar to [Kubernetes job controller](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/job/job_controller.go), which is a internal resource controller. In Kubernetes, a controller is a control loop that watches the shared state of the cluster through the apiserver and makes changes attempting to move the current state towards the desired state.
```
Comment (DjangoPeng) [inline docs/design_doc.md]: `an internal`

Comment (gaocegege) [reply docs/design_doc.md]: Fixed

File: docs/design_doc.md [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+    Replicas:  2
+    Template:
+      Metadata:
+        Creation Timestamp:  <nil>
+        Labels:
+          Index:                     1
+          Job _ Type:                PS
+          Kubeflow . Caicloud . Io:  true
+          Runtime _ Id:              nbbbg
+          Tf _ Job _ Name:           dist-training-job
```
Comment (Jimexist) [inline docs/design_doc.md]: why extra spaces

Comment (gaocegege) [reply docs/design_doc.md]: This is copied from `kubectl describe`..

Comment (gaocegege) [reply docs/design_doc.md]: I can file an issue for the problem.

**Comment by gaocegege** [comment]
@Jimexist Removed sensitive data now.

File: docs/design_doc.md [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+
+The controller initializes three resource informers: TFJob, Pod, and Service, and registers event handlers for the informers. Every time a pod/service/tfjob is created, updated, or deleted, the controller checks if actions should be taken to move the current state towards the desired state.
+
+A typical distributed TFJob contains one or more PS, one or more workers, and every PS/worker is run in a pod, and exposed to other PS/workers via a service:
+
+```
+TFJob
+  PS:
+    - ps-0
+      - po/ps-0
```
Comment (ScorpioCPH) [inline docs/design_doc.md]: nit: pod/ps-0? po is a little confused.

File: docs/design_doc.md [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+  workdir:
+    Type:  HostPath (bare host directory volume)
+    Path:  /home/ist/go/src/github.com/caicloud/kubeflow-controller/examples/workdir
+```
+
+Most of the configs are user-specified in [dist.yml](https://github.com/caicloud/kubeflow-controller/blob/master/examples/tfjob/dist.yml), and the labels and arguments are generated by the controller. The label is used to expose port, and the arguments is to hide the cluster spec from user code for ease of use.
+
+## Implementation
+
+Lifecycle management is the main responsibility, and the logic is in `pkg/controller/controller.go`. We keep a TTLCache of pod and service creates/deletes each TFJob expected to see, and updates the cache when the controller watches pods or services state changed.
```
Comment (ScorpioCPH) [inline docs/design_doc.md]: > We keep a TTLCache of pod and service creates/deletes each TFJob expected to see, and updates the cache when the controller watches pods or services state changed. 

Not very clear what we have done in the main logic.

Suggestion: We use `SharedInformers` to ListWatch the resource `TFJob`,  and register specify event handlers to fill the workqueue with `TFJob` creations, updates and deletions. Then process one item from the workqueue at a time:

- Parse to get pod-template, create pod if needed.
- Try to make the TFJobs's current state be more like the desired state.
- Do cleanup if the training completed.
- ...

Comment (gaocegege) [reply docs/design_doc.md]: SGTM

**Comment by gaocegege** [comment]
PTAL

File: docs/design_doc.md (hunk 2 above)
Comment (ScorpioCPH) [reply docs/design_doc.md]: Any update about this comment?

Comment (gaocegege-caicloud) [reply docs/design_doc.md]: #74

**Pull request merged**